packages/compass-aggregations/src/modules/pipeline-builder/pipeline-mode.ts

```diff
@@ -26,9 +34,28 @@ const reducer: Reducer<State> = (state = INITIAL_STATE, action) => {
   return state;
 };
 
-export const changePipelineMode = (mode: PipelineMode): PipelineModeToggledAction => ({
-  type: ActionTypes.PipelineModeToggled,
-  mode,
-});
+export const changePipelineMode = (
+  newMode: PipelineMode
+): PipelineBuilderThunkAction<void, PipelineModeToggledAction> => {
+  return (dispatch, getState, { pipelineBuilder }) => {
+    // Sync the PipelineBuilder
+    if (newMode === 'as-text') {
+      pipelineBuilder.stagesToSource();
+    } else {
+      pipelineBuilder.sourceToStages();
+    }
+
+    dispatch({
+      type: ActionTypes.PipelineModeToggled,
+      mode: newMode,
+      stageOperators: getStageOperatorsFromPipelineSource(getState(), pipelineBuilder),
```

Comment:
stageOperators are part of `as-text` slice and used to show stage list in toolbar.